File: web/di_core/LakeHouse/Domain/LakeJob/SQLJob.ts
import { LakeJob } from './LakeJob';
import { JobId, OrgId } from '@core/domain';
import { TimeScheduler } from '@/screens/DataIngestion/components/JobSchedulerForm/SchedulerTime/TimeScheduler';
import { LakeJobs } from '@core/LakeHouse/Domain/LakeJob/LakeJobs';
import { LakeJobStatus } from '@core/LakeHouse/Domain/LakeJob/LakeJobStatus';
import { LakeJobType } from '@core/LakeHouse/Domain/LakeJob/LakeJobType';
import { ResultOutput } from '@core/LakeHouse/Domain/LakeJob/OutputInfo/ResultOutput';
import { SchedulerOnce } from '@/screens/DataIngestion/components/JobSchedulerForm/SchedulerTime/SchedulerOnce';

export class SQLJob extends LakeJob {
  className = LakeJobs.SQL;
  jobType: LakeJobType = LakeJobType.SQL;
  query: string;
  outputs: ResultOutput[];

  constructor(
    orgId: OrgId,
    jobId: JobId,
    creatorId: string,
    name: string,
    lastRunTime: number,
    nextRunTime: number,
    lastSyncStatus: LakeJobStatus,
    currentJobStatus: LakeJobStatus,
    scheduleTime: TimeScheduler,
    query: string,
    outputs: ResultOutput[]
  ) {
    super(orgId, jobId, creatorId, name, lastRunTime, nextRunTime, lastSyncStatus, currentJobStatus, scheduleTime);
    this.query = query;
    this.outputs = outputs;
  }

  static create(query: string): SQLJob {
    return new SQLJob(
      LakeJob.DEFAULT_ID,
      LakeJob.DEFAULT_ID,
      '',
      '',
      0,
      0,
      LakeJobStatus.Initialized,
      LakeJobStatus.Initialized,
      new SchedulerOnce(Date.now()),
      query,
      []
    );
  }

  static fromObject(obj: any): SQLJob {
    return new SQLJob(
      obj.orgId,
      obj.jobId,
      obj.creatorId,
      obj.name,
      obj.lastRunTime,
      obj.nextRunTime,
      obj.lastRunStatus,
      obj.currentJobStatus,
      TimeScheduler.fromObject(obj.scheduleTime),
      obj.query,
      obj.outputs
    );
  }
}

File: job-scheduler/src/main/scala/datainsider/lakescheduler/domain/job/SqlJob.scala
package datainsider.lakescheduler.domain.job

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.module.scala.JsonScalaEnumeration
import datainsider.client.domain.scheduler.{NoneSchedule, ScheduleTime}
import datainsider.jobscheduler.domain.Ids.JobId
import datainsider.jobscheduler.util.JsonUtils
import datainsider.lakescheduler.domain.job.LakeJobStatus.LakeJobStatus
import datainsider.lakescheduler.domain.{HadoopResultOutput, LakeJobProgress, ResultOutput}
import datainsider.lakescheduler.domain.job.LakeJobType.LakeJobType

import java.sql.ResultSet

case class SqlJob(
    orgId: Long = -1L,
    jobId: JobId = 0,
    creatorId: String,
    name: String,
    lastRunTime: Long,
    nextRunTime: Long = System.currentTimeMillis(),
    @JsonScalaEnumeration(classOf[LakeJobStatusRef])
    lastRunStatus: LakeJobStatus,
    @JsonScalaEnumeration(classOf[LakeJobStatusRef])
    currentJobStatus: LakeJobStatus,
    scheduleTime: ScheduleTime = new NoneSchedule,
    yarnAppId: String = "",
    query: String,
    outputs: Seq[ResultOutput]
) extends LakeJob {

  override def jobType: LakeJobType = LakeJobType.Sql

  override def jobData: Map[String, Any] = {
    Map(
      "schedule_time" -> JsonUtils.toJson(scheduleTime),
      "query" -> query,
      "outputs" -> JsonUtils.toJson(outputs.toArray),
      "yarn_app_id" -> yarnAppId
    )
  }

  override def copyJobStatus(progress: LakeJobProgress): LakeJob = {
    progress.jobStatus match {
      case LakeJobStatus.Finished =>
        this.copy(
          currentJobStatus = LakeJobStatus.Finished,
          lastRunTime = progress.updatedTime,
          lastRunStatus = LakeJobStatus.Finished
        )
      case LakeJobStatus.Error =>
        this.copy(
          currentJobStatus = LakeJobStatus.Error,
          lastRunTime = progress.updatedTime,
          lastRunStatus = LakeJobStatus.Error
        )
      case _ =>
        this.copy(
          currentJobStatus = progress.jobStatus
        )
    }
  }

  override def customCopy(
      lastRunStatus: LakeJobStatus,
      currentJobStatus: LakeJobStatus,
      lastRunTime: JobId,
      creatorId: String,
      yarnAppId: String,
      nextRunTime: Long
  ): LakeJob =
    this.copy(
      currentJobStatus = currentJobStatus,
      lastRunStatus = lastRunStatus,
      lastRunTime = lastRunTime,
      creatorId = creatorId,
      yarnAppId = yarnAppId,
      nextRunTime = nextRunTime
    )
}

object SqlJob {
  def fromResultSet(rs: ResultSet): LakeJob = {
    val jobData: JsonNode = JsonUtils.readTree(rs.getString("job_data"))
    SqlJob(
      orgId = rs.getLong("organization_id"),
      jobId = rs.getLong("id"),
      name = rs.getString("name"),
      creatorId = rs.getString("creator_id"),
      lastRunTime = rs.getLong("last_run_time"),
      nextRunTime = rs.getLong("next_run_time"),
      lastRunStatus = LakeJobStatus.withName(rs.getString("last_run_status")),
      currentJobStatus = LakeJobStatus.withName(rs.getString("current_job_status")),
      scheduleTime = JsonUtils.fromJson[ScheduleTime](jobData.get("schedule_time").textValue()),
      query = jobData.get("query").textValue(),
      outputs = JsonUtils.fromJson[Seq[ResultOutput]](jobData.get("outputs").textValue()),
      yarnAppId = jobData.get("yarn_app_id").textValue()
    )
  }
}
